tests/test_main.py

```diff
@@ -2444,3 +2444,26 @@ def set_x(m: 'OuterModel') -> 'OuterModel':
     assert 'inner was here' in str(cs)
 
     assert OuterModel(inner=InnerModel()).x == 2
+
+
+def test_resolve_def_schema_from_core_schema() -> None:
+    class Inner(BaseModel):
+        x: int
+
+    class Marker:
+        def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+            schema = handler(source_type)
+            resolved = handler.resolve_ref_schema(schema)
```

Comment:
This used to fail because at the point when it was called `GenerateSchema` had no knowledge of the `Inner ` definition.